Add callback function to get JSON Output strings

davidBar-On opened the pull request:
* Version of iperf3 (or development branch, such as `master` or
  `3.1-STABLE`) to which this pull request applies:
master

* Issues fixed (if any): #1711

* Brief description of code changes (suitable for use as a commit message):
Add `iperf_set_test_json_callback()` that allows a user app to set a callback function to get the JSON output strings, instead of printing them to the output file.

swlars commented:
Thanks for the pull request! This looks useful!

bmah888 reviewed (commented):
Everything seems to work correctly. I have one question about the intended callback function...all of the other callbacks take a `test` argument. For those callbacks (such as `on_new_stream`, `on_test_start`, etc.), the `test` pointer gives all the information needed by the callback. The `json_callback` is different, in that it looks like it only takes a string containing a JSON rendered object. Do you think that having the `test` structure as well might be needed to give some context for working with the JSON? So then callback would take both a `(iperf_test *)` and a `(char *)`?

bmah888 commented:
Also, the code changes in #1805 that deal with existing callback hooks look relevant here as well.

davidBar-On commented:
> ... I'm just wondering if having a pointer to the test structure might be useful, say to refer to test parameters, etc.

Completely agree.  Added the `test` parameter and changed per #1805.  Now the callback setting function definition is:
`void iperf_set_test_json_callback(struct iperf_test *ipt, void (*callback)(struct iperf_test *, char *));`

bmah888 reviewed (approved):
This looks good, merging!
